apiserver/common/networkingcommon/networkconfigapi.go

```diff
@@ -0,0 +1,180 @@
+
+// The networkconfigapi package implements the network config parts
+// common to machiner and provisioner interface
+
+package networkingcommon
+
+import (
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/errors"
+	"github.com/juju/juju/apiserver/common"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/state"
+	"github.com/juju/juju/state/stateenvirons"
+)
+
+type NetworkConfigAPI struct {
+	st           *state.State
+	getCanModify common.GetAuthFunc
+	getAuthFunc  common.GetAuthFunc
+}
+
+func NewNetworkConfigAPI(st *state.State, getCanModify common.GetAuthFunc) *NetworkConfigAPI {
+	return &NetworkConfigAPI{
+		st:           st,
+		getCanModify: getCanModify,
+	}
+}
+
+func (api *NetworkConfigAPI) getMachine(tag names.Tag) (*state.Machine, error) {
+	entity, err := api.st.FindEntity(tag)
+	if err != nil {
+		return nil, err
+	}
+	return entity.(*state.Machine), nil
+}
+
+func (api *NetworkConfigAPI) getOneMachineProviderNetworkConfig(m *state.Machine) ([]params.NetworkConfig, error) {
+	instId, err := m.InstanceId()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	netEnviron, err := NetworkingEnvironFromModelConfig(
+		stateenvirons.EnvironConfigGetter{api.st},
+	)
+	if errors.IsNotSupported(err) {
+		logger.Infof("provider network config not supported: %v", err)
+		return nil, nil
+	} else if err != nil {
+		return nil, errors.Annotate(err, "cannot get provider network config")
+	}
+
+	interfaceInfos, err := netEnviron.NetworkInterfaces(instId)
+	if err != nil {
+		return nil, errors.Annotatef(err, "cannot get network interfaces of %q", instId)
+	}
+	if len(interfaceInfos) == 0 {
+		logger.Infof("not updating provider network config: no interfaces returned")
+		return nil, nil
+	}
+
+	providerConfig := NetworkConfigFromInterfaceInfo(interfaceInfos)
+	logger.Tracef("provider network config instance %q: %+v", instId, providerConfig)
+
+	return providerConfig, nil
+}
+
+func (api *NetworkConfigAPI) setOneMachineNetworkConfig(m *state.Machine, networkConfig []params.NetworkConfig) error {
+	devicesArgs, devicesAddrs := NetworkConfigsToStateArgs(networkConfig)
+
+	logger.Debugf("setting devices: %+v", devicesArgs)
+	if err := m.SetParentLinkLayerDevicesBeforeTheirChildren(devicesArgs); err != nil {
+		return errors.Trace(err)
+	}
+
+	logger.Debugf("setting addresses: %+v", devicesAddrs)
+	if err := m.SetDevicesAddressesIdempotently(devicesAddrs); err != nil {
+		return errors.Trace(err)
+	}
+
+	logger.Debugf("updated machine %q network config", m.Id())
+	return nil
+}
+
+func (api *NetworkConfigAPI) getMachineForSettingNetworkConfig(machineTag string) (*state.Machine, error) {
+	canModify, err := api.getCanModify()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	tag, err := names.ParseMachineTag(machineTag)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if !canModify(tag) {
+		return nil, errors.Trace(common.ErrPerm)
+	}
+
+	m, err := api.getMachine(tag)
```

Comment:
names.ParseMachineTag does return a names.MachineTag object, so we can have getMachine take one of those directly and feel more comfortable that FindEntity won't ever be called with something that isn't a Machine.